benchmarks/test_tcp_throughput.py

```diff
@@ -0,0 +1,75 @@
+from twisted.internet import reactor
+from twisted.internet.defer import Deferred
+from twisted.internet.endpoints import TCP4ClientEndpoint
+from twisted.internet.protocol import Factory, Protocol, ServerFactory
+from twisted.internet.testing import _benchmarkWithReactor as benchmarkWithReactor
+from twisted.protocols.wire import Echo
+
+
+class Server(Echo):
+    def connectionMade(self):
+        self.transport.setTcpNoDelay(True)
+
+
+class Counter(Protocol):
+    count = 0
+    sendAmount = 0
+    finished: None | Deferred = None
+
+    def connectionMade(self):
+        self.transport.setTcpNoDelay(True)
+
+    def dataReceived(self, b):
+        self.count += len(b)
+        if self.count == self.sendAmount:
+            self.transport.loseConnection()
+
+    def connectionLost(self, reason):
+        assert self.finished is not None
+        self.finished.callback(self.count)
+
+
+class Client(object):
+    def __init__(self, reactor, server):
+        self._reactor = reactor
+        self._server = server
+        self._sent = 0
+
+    def run(self, sendAmount, chunkSize):
+        self._sendAmount = sendAmount
+        self._bytes = b"x" * chunkSize
+        # Set up a connection
+        factory = Factory()
+        factory.protocol = Counter
+        d = self._server.connect(factory)
+        d.addCallback(self._connected)
+        return d
+
+    def _connected(self, client):
+        self._client = client
+        self._client.finished = Deferred()
+        self._client.sendAmount = self._sendAmount
+        client.transport.registerProducer(self, False)
+        return self._client.finished
+
+    def resumeProducing(self):
+        self._client.transport.write(self._bytes)
+        self._sent += len(self._bytes)
+        if self._sent == self._sendAmount:
+            self._client.transport.unregisterProducer()
+
+
+@benchmarkWithReactor
+async def test_tcp_throughput():
+    chunkSize = 16384
+    sendAmount = 1024 * 1024
+
+    server = ServerFactory()
+    server.protocol = Server
+    port = reactor.listenTCP(0, server)
```

Comment:
I am not sure if we are benchmarking client-side or server-side or both here.

-------

I was thinking that for the benchmark for client-side tests , we can make an exception and have a single server that is used by all the tests runs.

That is, create the server as a global/singleton/fixture outside of the benchmark run.

Reply:
It's benchmarking both I guess?

Reply:
The client send sends in 16KB chunks, the server-side likely responds in bigger chunks, and this also benchmarks reading... insofar as the goal isn't to measure speed of sending, but to measure speed of Twisted in these sort of scenarios, it seems OK as is?

Reply:
It's ok to benchmark both.

I guess that we can refactor this in the furure.

I was thinking that we might want to have separate benchmark tests for data processing speed for:
* client send
* client receive
* server send
* server receive


The current test is good enought to server as a general regression test.

I was thinking that for example if we improve performance for client send, while server receive performance is worse... the tests might still be ok, as the overall tests speed are not changed

Now... this is unlikely to happen